File: leetcode/20.valid_Parentheses.rb
=begin
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
=end

#Original
def is_valid1(s)
  parentheses = '{[()]}'
  str = '' 
  s.each_char do |char|
    char_idx = parentheses.index(char)
    next unless char_idx
    if str[-1] && parentheses.index(str[-1]) + char_idx == 5
      str = str[0...-1]
    else
      str += char
    end
  end
  str.empty? 

end

#Revised
def is_valid(s)
  str = [] 
  s.each_char do |char|
    case char
    when "[", "{", "(" 
      str.push(char)
    when "]"
      return false unless str.pop == "["
    when "}"
      return false unless str.pop == "{"
    when ")"
      return false unless str.pop == "("
    end
  end
  str.empty?
end

File: leetcode/pythonSolutions/20.valid_parentheses.py
'''
20. Valid Parentheses

Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true
Example 2:

Input: "()[]{}"
Output: true
Example 3:

Input: "(]"
Output: false
Example 4:

Input: "([)]"
Output: false
Example 5:

Input: "{[]}"
Output: true
'''

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        list = []
        length = len(s)
        for i in range(length):
          char = s[i]
          if char in ['(','[','{']:
            list.append(char)
          else:
            if len(list) < 1: return False
            lastChar = list.pop()
            if (char == ')') and (lastChar != '('):  return False
            if char == ']' and lastChar != '[': return False 
            if char == '}' and lastChar != '{': return False
        return len(list) == 0

str = "{[]}"
sol = Solution()
print(sol.isValid(str))

        